language/move-vm/types/src/values/values_impl.rs

```diff
@@ -302,33 +286,104 @@ impl Value {
                     && v.len() == 1
                     && matches!(&v[0], ValueImpl::Address(_))
             }
-            _ => false,
+            _ => self.0.is_primitive_or_vector_arg(sig),
         }
     }
 
+    // Check that a `Value` is either a possible constant or a `Signer`.
+    // This function only filters obvious bad args. It is possible to fool
+    // this function to take inconsistent `Vector`s.
+    // `is_valid_arg` above makes the consistency check against the signature
+    // of the function invoked.
     pub fn is_constant_or_signer_ref(&self) -> bool {
         match &self.0 {
+            ValueImpl::ContainerRef(ContainerRef::Local(Container::StructR(r))) => {
+                let v = &*r.borrow();
+                v.len() == 1 && matches!(&v[0], ValueImpl::Address(_))
+            }
+            _ => self.0.is_constant(),
+        }
+    }
+}
+
+impl ValueImpl {
+    fn is_primitive_or_vector_arg(&self, sig: &SignatureToken) -> bool {
+        match (sig, &self) {
+            (SignatureToken::U8, ValueImpl::U8(_))
+            | (SignatureToken::U64, ValueImpl::U64(_))
+            | (SignatureToken::U128, ValueImpl::U128(_))
+            | (SignatureToken::Bool, ValueImpl::Bool(_))
+            | (SignatureToken::Address, ValueImpl::Address(_)) => true,
+            (SignatureToken::Vector(ty), ValueImpl::Container(c)) => match (&**ty, c) {
+                (SignatureToken::Bool, Container::VecBool(_))
+                | (SignatureToken::U8, Container::VecU8(_))
+                | (SignatureToken::U64, Container::VecU64(_))
+                | (SignatureToken::U128, Container::VecU128(_))
+                | (SignatureToken::Address, Container::VecAddress(_)) => true,
+                (SignatureToken::Vector(inner), Container::VecC(values)) => {
+                    if !Self::is_signature_constant(inner) {
+                        return false;
+                    }
+                    for value in &*values.borrow() {
+                        if !value.is_primitive_or_vector_arg(ty) {
+                            return false;
+                        }
+                    }
+                    true
+                }
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    // Walks a `SignatureToken` to verify that represents a constant
+    fn is_signature_constant(sig: &SignatureToken) -> bool {
+        match sig {
+            SignatureToken::U8
+            | SignatureToken::U64
+            | SignatureToken::U128
+            | SignatureToken::Bool
+            | SignatureToken::Address => true,
+            SignatureToken::Vector(ty) => match &**ty {
```

Comment:
Same thing here: Recurse earlier so you don't have to repeat the match

Reply:
+1